### DISCUSSION TITLE
Adhere All Zustand State To a Localized Scope

### DISCUSSION THREAD

**Discussion by Jonathynlee** [pmndrs/zustand#2177] [Q&A] [answered]
Right now I am using Zustand to keep track of all of my state in a react-three-fiber application. In the application I rollup specific components into useable modular 3D components (and web components). As designed by Zustand (I think), when components are used that share the same state, the state gets updated in both places. I want to prevent this from happening and would like each component to keep track of its own state only. Is it possible to localize the state of each individual component?

I did quite bit of search for a solution on this and was not able to find a concrete answer that fits my use case. 

How it works today:
![image](https://github.com/pmndrs/zustand/assets/54726329/e2d1e687-d286-4368-aa18-bfc98497f2a2)

How I would like it to work:
![image](https://github.com/pmndrs/zustand/assets/54726329/8e8fc9fb-2f16-4121-aa4b-736977701b98)

If there was a way to do this with rollup, I think that would be an even better method. Thanks ahead of time!

**Comment by dai-shi** [comment]
Not sure if I get the point, but does React Context help?

**Reply by Jonathynlee** [reply]
The purpose is to gain all of the benefits of Zustand state management but essentially creating zustand scopes where only a component and all of its children can access the state in question. 

React Context is part of the solution but I would need a way to pass the zustand context into the react context.

**Reply by dai-shi** [reply]
Sorry for the lack of details. I mean to use React Context and Zustand store.

**Reply by Jonathynlee** [reply]
That sounds like a good solution but I haven't found an effective way to implement it. Specifically using typescript and combining serveral stores?.  

As a side note, due to R3F's reliance on zustand 3, zustand 3 is being used in this project. 

Do you have any examples of this implementation? 

To clarify further, imagine we have a series of many stores being used, I need a way to pass only the stores pertinent to the component is question to React Context Provider

**Reply by dbritto-dev** [reply]
@Jonathynlee what about using shallow?

**Reply by Jonathynlee** [reply]
Could you please elaborate what you mean by 'shallow' @dbritto-dev ?

**Reply by dbritto-dev** [reply]
@Jonathynlee https://docs.pmnd.rs/zustand/guides/prevent-rerenders-with-use-shallow

**Reply by Jonathynlee** [reply]
@dbritto-dev I am not sure how useShallow relates to my issue. From my understanding this just prevents rerenders when a state changes but a desired output does not change. Please correct me if I am wrong.

I am trying to use the same defined state context with different component where each individual component can manage it own version of that same state context.

**Reply by dbritto-dev** [reply]
@Jonathynlee in that case a closure and useRef would be enough

```ts
const createUseCounter = () => {
  return create((set) => ({
    count: 0,
    inc: () => set((state) => ({ count: state.count + 1 })),
  }));
}
```

```tsx
const Counter = () => {
  const useCounterRef = useRef()
  if (!useCounterRef.current) {
    useCounterRef.current = createUseCounter()
  }
  const { count, inc } = useCounterRef.current()

  return <button type="button" onClick={inc}>{count}</button>
}
```

**Reply by Jonathynlee** [reply]
@dbritto-dev  this sounds promising. I have never heard of a closure. Do you have a link to some documentation on it? Did a quick google search and nothing came up about it immediately.

**Reply by Jonathynlee** [reply]
@dbritto-dev in my real scenario my components have many children in which my different states affect those children differently. I also have many different contexts that already exist. In the solution you have. In order to do what you you suggested would I either need to props pass or use react context provider?

In my real scenario, I would like it to work similarly to how React three fiber does works. You can create a canvas which has its own set of state that it manages itself, even though is has many children. I did see mention of useBoundsState but wasn't sure if I was digging in the right rabbit hole. 

Any thougths?

**Reply by dbritto-dev** [reply]
@Jonathynlee yes, you can combine that with useContext + Context API

**Reply by dbritto-dev** [reply, marked as answer]
@Jonathynlee here you go -> https://codesandbox.io/p/sandbox/gifted-dan-m2f8zj?file=%2Fsrc%2FApp.tsx%3A29%2C32

**Reply by Jonathynlee** [reply]
Ill try it with my code base and see if I can get it working. Thank you!

**Reply by Jonathynlee** [reply]
@dbritto-dev  I had to set up a series of wrappers to achieve the desire output. 

It would be incredible in the future if there was a way to set a zustand container where  all state used within would be bound to just that context. Also, because it is a waterfall, states only have access to the zustand states above them. I will need to combine the states at some point, somehow.

**Reply by dbritto-dev** [reply]
@Jonathynlee what you mean? do you have any example?

**Reply by Jonathynlee** [reply]
@dbritto-dev 
I have several stores that exist in my code.  13 to be exact. With the methodology provided, I will be stacking those stores to provide all stores needed.

Using the image below, it can be assumed that:
Store A will no have access to any other store
Store B will only have access to Store A
![image](https://github.com/pmndrs/zustand/assets/54726329/0f152b2f-4eb8-4de8-909c-bfcfca2c6db2)

**Reply by Jonathynlee** [reply]
@dbritto-dev 
Additionally
For the zustand container concept, this was my thought. If there was a way to create a boundary around some react components that would maintain the state within that boundary, this would solve my problem as well. 
Here is an image to describe what I mean. 
![image](https://github.com/pmndrs/zustand/assets/54726329/5294b4e5-ccac-4ea4-b8ef-2f36bdf41034)

**Reply by dbritto-dev** [reply]
@Jonathynlee yes, you can achieve that. BTW, you can create a component provides Store A, Store B and Store C

```tsx
const UseStoreAContext = createContext()
const UseStoreBContext = createContext()
const UseStoreCContext = createContext()

const BoundaryStoreProvider = ({ children }) => {
  const useStoreARef = useRef<UseBoundStore<StoreApi<StoreA>>>();
  if (!useStoreARef.current) {
    useStoreARef.current = createUseStoreA();
  }
  const useStoreBRef = useRef<UseBoundStore<StoreApi<StoreB>>>();
  if (!useStoreBRef.current) {
    useStoreBRef.current = createUseStoreB();
  }
  const useStoreCRef = useRef<UseBoundStore<StoreApi<StoreC>>>();
  if (!useStoreCRef.current) {
    useStoreCRef.current = createUseStoreC();
  }
  
  return (
    <UseStoreAContext.Provider value={useStoreARef.current}>
      <UseStoreBContext.Provider value={useStoreBRef.current}>
        <UseStoreCContext.Provider value={useStoreCRef.current}>{children}</UseStoreCContext.Provider>
      </UseStoreBContext.Provider>
    </UseStoreAContext.Provider>
  );
}
```

**Reply by Jonathynlee** [reply]
This will cause the issue where useStoreARef is unable to access useStoreBRef, correct?

**Reply by dbritto-dev** [reply]
@Jonathynlee it depends on how you order each context provider

**Reply by Jonathynlee** [reply]
Right, so if it is possible it would be really awesome if I could leave all my states as they were and add some container wrapper that would allow all of those stores to communicate to each other while allowing me to componentize. 

Do you know if it possible to combine all of my stores to be used as a single provider around my component? @dbritto-dev 
Or is providing 13 Providers the only way?

**Reply by dbritto-dev** [reply]
@Jonathynlee we have don't a way to combine stores, maybe slices pattern would be good for your case

**Comment by oliveryasuna** [comment]
@dbritto-dev The CodeSandbox gives 404.

**Reply by dbritto-dev** [reply]
@oliveryasuna  yeah, we are moving away from codesandbox to stackblitz but we lost some demos

**Reply by dbritto-dev** [reply]
so far I created this collection -> https://stackblitz.com/@dbritto-dev/collections/zustand
